src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,214 @@
+| (n+1) := circulant_mul
+
+/-- Multiplication of circulant matrices commutes when the elements do. -/
+lemma circulant_mul_comm
+  [comm_semigroup α] [add_comm_monoid α] [fintype n] [add_comm_group n] (v w : n → α) :
+  circulant v ⬝ circulant w = circulant w ⬝ circulant v :=
+begin
+  ext i j,
+  simp only [mul_apply, circulant, mul_comm],
+  refine fintype.sum_equiv ((equiv.sub_left i).trans (equiv.add_right j)) _ _ _,
+  intro x,
+  congr' 2,
+  { simp },
+  { simp only [equiv.coe_add_right, function.comp_app,
+               equiv.coe_trans, equiv.sub_left_apply],
+    abel }
+end
+
+lemma fin.circulant_mul_comm [comm_semigroup α] [add_comm_monoid α] :
+  ∀ {n} (v w : fin n → α), circulant v ⬝ circulant w = circulant w ⬝ circulant v
+| 0     := dec_trivial
+| (n+1) := circulant_mul_comm
+
+/-- `k • circulant v` is another circulantcluant matrix `circulant (k • v)`. -/
+lemma circulant_smul [has_sub n] [has_scalar R α] (k : R) (v : n → α) :
+  circulant (k • v) = k • circulant v :=
+by ext; simp
+
+@[simp] lemma circulant_single_one
+  (α n) [has_zero α] [has_one α] [decidable_eq n] [add_group n] :
+  circulant (pi.single 0 1 : n → α) = (1 : matrix n n α) :=
+by { ext i j, simp [one_apply, pi.single_apply, sub_eq_zero] }
+
+@[simp] lemma circulant_single
+  (n) [semiring α] [decidable_eq n] [add_group n] [fintype n] (a : α) :
+  circulant (pi.single 0 a : n → α) = scalar n a :=
+begin
+  ext i j,
+  simp [pi.single_apply, one_apply, sub_eq_zero],
+end
+
+/-- Note we use `↑i = 0` instead of `i = 0` as `fin 0` has no `0`.
+    This means that we cannot state this with `pi.single`
+    as we did with `matrix.circulant_single`. -/
+lemma fin.circulant_ite (α) [has_zero α] [has_one α] :
+  ∀ n, circulant (λ i, ite (↑i = 0) 1 0 : fin n → α) = 1
+| 0     := dec_trivial
+| (n+1) :=
+begin
+  rw [←circulant_single_one],
+  congr' with j,
+  simp only [pi.single_apply, fin.ext_iff],
+  congr
+end
+
+/-- The circulant matrix `circulant v` is symmetric iff
+    `∀ i j, v (j - i) = v (i - j)`. -/
+lemma circulant_is_symm_iff' [has_sub n] {v : n → α} :
+  (circulant v).is_symm ↔ ∀ i j, v (j - i) = v (i - j) :=
+by simp [is_symm.ext_iff, circulant]
+
+/-- The circulant matrix `circulant v` is symmetric iff
+    `v (- i) = v i` if `[add_group n]`. -/
+lemma circulant_is_symm_iff [add_group n] {v : n → α} :
+  (circulant v).is_symm ↔ ∀ i, v (- i) = v i :=
+begin
+  rw [circulant_is_symm_iff'],
+  split,
+  { intros h i, convert h i 0; simp },
+  { intros h i j, convert h (i - j), simp }
+end
+
+lemma fin.circulant_is_symm_iff :
+  ∀ {n} {v : fin n → α}, (circulant v).is_symm ↔ ∀ i, v (- i) = v i
+| 0     := λ v, by simp only [circulant_is_symm_iff', is_empty.forall_iff]
+| (n+1) := λ v, circulant_is_symm_iff
+
+/-- If `circulant v` is symmetric, `∀ i j : I, v (j - i) = v (i - j)`. -/
+lemma circulant_is_symm_apply'
+  [has_sub n] {v : n → α} (h : (circulant v).is_symm) (i j : n) :
+  v (j - i) = v (i - j) :=
+circulant_is_symm_iff'.1 h i j
+
```

Comment:
I also don't think this one is worth having, since the proof is just `h.apply _ _`.

Reply:
Right. We have this API for `symm` now.